File: src/main/scala/com/gitlab/dhorman/cryptotrader/core/Market.scala
package com.gitlab.dhorman.cryptotrader.core

import io.circe.{Decoder, Encoder, KeyDecoder, KeyEncoder}

case class Market(baseCurrency: Currency, quoteCurrency: Currency) {
  def b: Currency = baseCurrency
  def q: Currency = quoteCurrency

  def tpe(currency: Currency): Option[CurrencyType] = {
    if (currency == baseCurrency) return Some(CurrencyType.Base)
    if (currency == quoteCurrency) return Some(CurrencyType.Quote)
    None
  }

  def currency(tpe: CurrencyType): Currency = tpe match {
    case CurrencyType.Base => baseCurrency
    case CurrencyType.Quote => quoteCurrency
  }

  def other(currency: Currency): Option[Currency] = {
    if (currency == baseCurrency) return Some(quoteCurrency)
    if (currency == quoteCurrency) return Some(baseCurrency)
    None
  }

  def orderType(targetCurrency: Currency): Option[OrderType] = {
    tpe(targetCurrency) map orderType
  }

  def orderType(currencyType: CurrencyType): OrderType = currencyType match {
    case CurrencyType.Base => OrderType.Sell
    case CurrencyType.Quote => OrderType.Buy
  }

  def contains(currency: Currency): Boolean = {
    currency == baseCurrency || currency == quoteCurrency
  }

  def find(currencies: Iterable[Currency]): Option[Currency] = {
    currencies.find(currency => currency == baseCurrency || currency == quoteCurrency)
  }

  override def toString: Currency = s"${baseCurrency}_$quoteCurrency"
}

object Market {
  implicit val encoder: Encoder[Market] = Encoder.encodeString.contramap(_.toString)

  implicit val decoder: Decoder[Market] = Decoder.decodeString.emap(market => {
    val currencies = market.split('_')

    if (currencies.length == 2) {
      Right(Market(currencies(0), currencies(1)))
    } else {
      Left(s"""Market "$market" not recognized""")
    }
  })

  implicit val keyEncoder: KeyEncoder[Market] = KeyEncoder.encodeKeyString.contramap(_.toString)
  implicit val keyDecoder: KeyDecoder[Market] = KeyDecoder.decodeKeyString.map(convert)

  implicit def convert(markets: Iterable[String]): Iterable[Market] = {
    markets.map(_.split('_')).filter(_.length == 2).map(c => Market(c(0), c(1)))
  }

  implicit def convert(market: String): Market = {
    val currencies = market.split('_')
    require(currencies.length == 2, "market must be in format A_B.")
    Market(currencies(0), currencies(1))
  }
}

File: src/main/kotlin/com/gitlab/dhorman/cryptotrader/core/Market.kt
package com.gitlab.dhorman.cryptotrader.core

import com.fasterxml.jackson.core.JsonGenerator
import com.fasterxml.jackson.core.JsonParser
import com.fasterxml.jackson.core.JsonToken
import com.fasterxml.jackson.databind.*
import com.fasterxml.jackson.databind.annotation.JsonDeserialize
import com.fasterxml.jackson.databind.annotation.JsonSerialize
import com.gitlab.dhorman.cryptotrader.service.poloniex.model.Currency
import com.gitlab.dhorman.cryptotrader.service.poloniex.model.CurrencyType
import com.gitlab.dhorman.cryptotrader.service.poloniex.model.OrderType

@JsonSerialize(using = Market.Companion.Encoder::class)
@JsonDeserialize(using = Market.Companion.Decoder::class)
data class Market(val baseCurrency: Currency, val quoteCurrency: Currency) {
    val b get() = baseCurrency
    val q get() = quoteCurrency

    fun tpe(currency: Currency): CurrencyType? {
        if (currency == baseCurrency) return CurrencyType.Base
        if (currency == quoteCurrency) return CurrencyType.Quote
        return null
    }

    fun currency(tpe: CurrencyType): Currency {
        return when (tpe) {
            CurrencyType.Base -> baseCurrency
            CurrencyType.Quote -> quoteCurrency
        }
    }

    fun other(currency: Currency): Currency? {
        if (currency == baseCurrency) return quoteCurrency
        if (currency == quoteCurrency) return baseCurrency
        return null
    }

    fun orderType(targetCurrency: Currency): OrderType? {
        val currencyType = tpe(targetCurrency)

        return if (currencyType != null) {
            orderType(currencyType)
        } else {
            null
        }
    }

    fun orderType(targetCurrencyType: CurrencyType): OrderType {
        return when (targetCurrencyType) {
            CurrencyType.Base -> OrderType.Sell
            CurrencyType.Quote -> OrderType.Buy
        }
    }

    fun targetCurrency(orderType: OrderType): Currency {
        return when (orderType) {
            OrderType.Sell -> baseCurrency
            OrderType.Buy -> quoteCurrency
        }
    }

    fun contains(currency: Currency): Boolean {
        return currency == baseCurrency || currency == quoteCurrency
    }

    fun find(currencies: Iterable<Currency>): Currency? {
        return currencies.find { currency -> currency == baseCurrency || currency == quoteCurrency }
    }

    override fun toString(): String = "${baseCurrency}_$quoteCurrency"

    companion object {
        class Encoder : JsonSerializer<Market>() {
            override fun serialize(value: Market, gen: JsonGenerator, serializers: SerializerProvider) {
                gen.writeString(value.toString())
            }
        }

        class Decoder : JsonDeserializer<Market>() {
            override fun deserialize(p: JsonParser, ctx: DeserializationContext): Market {
                if (p.currentToken() == JsonToken.VALUE_STRING) {
                    return p.valueAsString.toMarket()
                } else {
                    throw JsonMappingException(ctx.parser, "Can't parse market")
                }
            }
        }

        class KeyDecoder : KeyDeserializer() {
            override fun deserializeKey(key: String?, ctxt: DeserializationContext?): Any {
                return key?.toMarket() ?: throw Exception("Key can't be null")
            }
        }

        fun String.toMarket(): Market {
            val currencies = this.split('_')

            if (currencies.size == 2) {
                return Market(currencies[0], currencies[1])
            } else {
                throw Exception("""Market "$this" not recognized""")
            }
        }
    }
}
